### ISSUE TITLE
New: rule no-magic-number

### ISSUE THREAD

**Issue by kombucha** [eslint/eslint#4027]
**When does this rule warn? Please describe and show example code:**

This rule will warn when magic numbers are used in the code. For example :
```javascript
// Valid code
var TAX_PERCENTAGE = 0.25;

var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * TAX_PERCENTAGE);
```

```javascript
// Invalid code : Warns "No magic number: 0.25"
var dutyFreePrice = 100,
    finalPrice = dutyFreePrice + (dutyFreePrice * 0.25); 
```

The configuration takes 3 parameters :

* `ignore`: An array of numbers to ignore. It's set to `[0, 1, 2]` by default.
* `detectObjects`: A boolean to specify if we should detect numbers when setting object properties for example. `false` by default.
* `enforceConst`: A boolean to specify if we should check for the const keyword in variable declaration of numbers. `false` by default.

**Is this rule preventing an error or is it stylistic?**

It's more of a best practice / stylistic rule. Avoiding 'magic numbers' is good for understanding the code and refactoring it.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

I think it's a good candidate for inclusion because its application is really broad. "Magic numbers" and their usage is not something specific to my team, but concerns every developers (it even has its own [wikipedia article](https://en.wikipedia.org/wiki/Magic_number_(programming)) :) ).

**Are you willing to create the rule yourself?**

Yes :). Or to be more precise, I used @danielstjules's library, [buddy.js](https://github.com/danielstjules/buddy.js/), as the basis for this rule (code and specifications). I liked what buddy.js did, but I didn't want to use yet another tool for checking for magic numbers.
I will make it a plugin if your think it's not worth integrating it in core of course.

Cheers

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by lo1tuma** [comment]
Sounds good to me. 
This was also mentioned in the webstorm rollup #667

**Comment by ilyavolodin** [comment]
Agree. It's a good rule to have.

**Comment by danielstjules** [comment]
Nice! :)

Is it compatible with buddy.js for all the fixtures here https://github.com/danielstjules/buddy.js/tree/master/spec/fixtures ?

**Comment by kombucha** [comment]
Should be, yeah !
